### PR TITLE
Added FanOut/FanIn Pattern

### PR DESCRIPTION
**Pull request by karthikbhat13**
FanOut/FanIn Pattern

- The pattern refers to executing multiple instances of the activity function on multiple chunks in parallel
- Resolves iluwatar#1627

Pull request description
- Added code that implements the design pattern and test cases

### CODE REVIEW

File: fanout-fanin/README.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
+---
+layout: pattern
+title: FanOut-FanIn
```
Comment (iluwatar) [inline fanout-fanin/README.md:3]:
```suggestion
title: Fan-Out/Fan-In 
```

File: fanout-fanin/README.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
+---
+layout: pattern
+title: FanOut-FanIn 
+folder: fanout-fanin
+permalink: /patterns/fanout-fanin/
+categories: Concurrency
```
Comment (iluwatar) [inline fanout-fanin/README.md:6]:
```suggestion
categories: Integration
```

File: fanout-fanin/README.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+language: en
+tags:
+- Cloud distributed
+- Performance
```
Comment (iluwatar) [inline fanout-fanin/README.md:10]:
```suggestion
- Microservices
```

File: fanout-fanin/README.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+---
+
+## Intent
+The pattern that refers to executing multiple instances of the activity function concurrently. Generally "FanOut" 
+process is splitting the data into multiple chunks that would be passed to the activity function. "FanIn" process is 
+basically aggregating the result from individual instance of activity function and forming a single final result. <br />
+
+Each activity function is a long-running process that takes in a chunk of data. A callback function, or a Consumer 
+is passed to the activity function, that will be called when the long-running process gets over. The consumer will 
+process the individual return value from activity function as they are done running into a final result.
```
Comment (iluwatar) [inline fanout-fanin/README.md:20]: The term activity function is not self-explanatory. Could we perhaps talk about microservices?

Comment (iluwatar) [reply fanout-fanin/README.md:20]: We should also explain how the caller eventually gets the aggregated result.

Comment (iluwatar) [reply fanout-fanin/README.md:20]: We need an explanation section and some of the details could be placed there.

File: fanout-fanin/README.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+## Class diagram
+![alt-text](./etc/fanout-fanin.png)
+
+## Applicability
+
+Use this pattern when you can chunk the workload or load into multiple chunks that can be dealt with separately.
+
```
Comment (iluwatar) [inline fanout-fanin/README.md:28]: I think we could add a related patterns section, see https://github.com/iluwatar/java-design-patterns/wiki/02.-Pattern-template. Some of the microservices patterns we already have seemed to be quite similar.

**Review by iluwatar — CHANGES_REQUESTED** [review]
Thanks for implementing this pattern. Looks good in general. I added some improvement requests, please take a look and let us know when you're ready for another review.

**Comment by iluwatar** [comment]
@karthikbhat13 I'm not sure what has happened, but the PR doesn't have any changed files anymore. Can you please take a look at it?

**Comment by karthikbhat13** [comment]
Oh, didn't notice that. Lemme check. There is a hook to auto pull master branch to base repo. It probably replaced my branch. I will update again.

**Comment by karthikbhat13** [comment]
@iluwatar Can I update this PR only to have the changes again? I can't figure it out. For now, I created one more PR, https://github.com/iluwatar/java-design-patterns/pull/1799

Let me know what I can do. Thanks

**Comment by iluwatar** [comment]
@karthikbhat13 either way, but please ensure that you have addressed my earlier review comments.

**Comment by karthikbhat13** [comment]
> @karthikbhat13 either way, but please ensure that you have addressed my earlier review comments.

@iluwatar Have done that. I have fixed the comments and added the explanation and related patterns sections. https://github.com/iluwatar/java-design-patterns/pull/1799/commits/1948342f4f423cbeba5bedac9efa04ffc7168d02 of https://github.com/iluwatar/java-design-patterns/pull/1799

**Comment by iluwatar** [comment]
Ok, let's continue the process in #1799.

**Comment by karthikbhat13** [comment]
> Ok, let's continue the process in #1799.

@iluwatar There was the issue of pull bot again in this PR too. I have create a new one with custom branch,  https://github.com/iluwatar/java-design-patterns/pull/1800. Please consider this whenever convenient.

**Pull request merged**